Basic aim at target implementation

linglejack06 opened the pull request:
- utilizes basic command to call navigation method aim at target (gets yaw speed from camera best target in update inputs)
- then calls arcade drive from command with 0 forward speed and specified rotation speed

linglejack06 committed:
add basic ability to aim at target

- utilizes basic command to call navigation method aim at target (gets yaw speed from camera best target in update inputs)
- then calls arcade drive from command with 0 forward speed and specified rotation speed

linglejack06 commented:
Will this work? confused on how it will get to specified yaw when it will continue to set same speed ( do i need a pid controller or just a new method in drive for it?)

honzikschenk committed:
Update issue templates (#28)

jkleiber commented:
The concept I would recommend would be to establish a state machine in the DriveSubsystem with Actions that the user can take (it might be good to look at the discussion I had with Honzik in https://github.com/team401/2024-Robot-Code/pull/23)

Basically in your aim command you would tell the drivetrain you would like to take an AIM action, and then the drivetrain would have a PID for controlling the yaw to match a target yaw. The DriveSubsystem would need a Pose supplier from the Navigation subsystem and could use that to compute the target yaw. 

In drivetrain periodic you would be constantly controlling the drivetrain based on the mode. I recommend doing something similar to what Honzik did in that PR linked above since it will be scalable and easy to maintain. I would also recommend creating a state machine google sheet for everybot like this one for the main bot: https://docs.google.com/spreadsheets/d/1UDniDztXg6ygjTsRVeG_5nA809lSZbLa9G5Es8bKwF4/edit#gid=0

jkleiber commented:
Just to add on real quick based on a quick look at this PR - I'd rather us do things based on your pose estimate instead of purely vision. If you did it only based on vision (or only based on odometry) then you lose out on the other information the navigation system is fusing together.

linglejack06 commented:
That makes sense, I'll try to incorporate pose estimation into it. As for the state machine, couldn't you also interrupt the default drive command with an aim command that would then pass a pose to drive subsystem?

linglejack06 committed:
add ability to toggle 3d field sim on through use of constants file (#25)

linglejack06 committed:
add ability to toggle 3d field sim on through use of constants file (#25)

jkleiber commented:
Yeah passing a pose through a command works. It still would be good to make a state machine in drive because then the drive control logic stays centralized in the subsystem

jkleiber committed:
#18: migrates main robot CI to everybot (#29)

linglejack06 committed:
add conditional to check if there is new frame to add

linglejack06 committed:
add conditional to check if there is new frame to add

linglejack06 committed:
format code

linglejack06 committed:
format code

linglejack06 committed:
add tranform 3d input in vision classes

linglejack06 committed:
add tranform 3d input in vision classes

linglejack06 committed:
get pose2d from pose estimation not just vision

linglejack06 committed:
get pose2d from pose estimation not just vision

linglejack06 committed:
remove transform from target

- in favor of using solely pose estimation with known target on field

linglejack06 committed:
remove transform from target

- in favor of using solely pose estimation with known target on field

linglejack06 committed:
move field layout to constants file

linglejack06 committed:
add desired target method to set pose of what we want to aim at

linglejack06 committed:
add desired target method to set pose of what we want to aim at

linglejack06 committed:
add heading error method to send to drivetrain

linglejack06 commented:
@jkleiber is this how you would calculate the heading error?

linglejack06 committed:
Merge branch 'main' into aim-at-target

linglejack06 committed:
fix import

jkleiber commented:
Current heading looks right, but you'd want to determine the target heading as the vector from the robot center to the target center.

I would recommend drawing that and seeing if you can come up with a way to find what that heading is (it would be a good piece of documentation to add to the repo as well!). As a hint, I would assume 0 degrees to be the long axis of the field

If you can't figure it out by Wednesday, lmk and I can

linglejack06 commented:
Ok, I'll try to figure that out. Thank you

linglejack06 commented:
am i essentially trying to find the angle between the robot vector and target vector?

jkleiber commented:
Yes, that's correct @linglejack06

linglejack06 committed:
add formula to calculate angle using dot product

linglejack06 committed:
format

linglejack06 commented:
Filled out what I think will work. Not positive though, as I don't have much experience working with vectors. Let me know how it looks.

linglejack06 commented:
I also saw a way using cross product formula with a z element as well ( would z be the heading? )

jkleiber commented:
@linglejack06 you are using a correct equation (acos(a dot b / norm(a) * norm(b)))

The inputs are just not correct. Here's a diagram that could help
![auto-aim](https://github.com/team401/2024-everybot/assets/10378677/9aa0cd33-ac88-4256-95e0-1539a7a34284)

For the equation you have, you would want to find the angle between the green and blue vectors. What you are currently finding is the angle between the robot's position vector and the speaker position vector. So you need to construct these two vectors based on your known information.

For the green vector, I would create a unit vector to keep things simple

linglejack06 commented:
Ok thank you, I'll try to update the inputs

linglejack06 committed:
convert current vector to unit vector magnitude

linglejack06 committed:
switch to use built in transforms?"

linglejack06 committed:
Merge branch 'main' into aim-at-target

linglejack06 committed:
add unit vector for robot

linglejack06 committed:
update vector calculations

linglejack06 committed:
add placeholder target id and update command

linglejack06 committed:
update target vector with robot pose x instead of unit vector x

linglejack06 commented:
@jkleiber this should be ready for review if I did all the math right.

jkleiber commented on src/main/java/frc/robot/subsystems/Navigation.java:
```diff
@@ -50,45 +53,79 @@ public Navigation (DoubleSupplier leftDistance, DoubleSupplier rightDistance, Su
+        double targetVectorY =
+                desiredTargetPose.getY() - poseEstimator.getEstimatedPosition().getY();
+
+        double dotProduct = (robotVectorX * targetVectorX) + (robotVectorY - targetVectorY);
```
this would be `robotVectorY * targetVectorY` in the second parenthesis

linglejack06 replied:
Yep that was a typo

jkleiber commented on src/main/java/frc/robot/commands/AimAtTarget.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public void initialize() {
+        navSubsystem.setDesiredTarget(targetId);
```
It still seems that this command is targeting a specific AprilTag rather than a position on the field

linglejack06 replied:
Yes, I think it's the pose of the target, but all targets should be centered above the speaker, etc. so shouldn't the angle still work?

jkleiber replied:
If we have passed in the speaker's AprilTag target ID based on our alliance color that could work

linglejack06 replied:
Yes that is what I was working on setting up in constants, so one could say constants.speakertag and it would be set to current field alliance

jkleiber replied:
Ideally the alliance color would be received from the DriverStation (there's a way to request the current alliance color) rather than being a hardcoded constant. This would need to run in RobotContainer similar to what Charlie has done in the main repo

linglejack06 replied:
ok ill update that now

jkleiber commented on src/main/java/frc/robot/commands/AimAtTarget.java:
```diff
@@ -0,0 +1,29 @@
+        this.navSubsystem = navSubsystem;
+        this.targetId = targetId;
+
+        addRequirements(driveSubsystem, navSubsystem);
```
The drivetrain is required, but not used in this command. We should probably get #30 merged so you can build on that

linglejack06 replied:
Yeah, I left that as I need a drive method to call with the heading error ( with a pid controller in drive? Or I could add the controller to navigation)

jkleiber replied:
The PID controller would be in drive and would be a new state (or maybe @AlexEvergarden has already made a placeholder state for this). That PR is approved so it can be merged whenever

linglejack06 replied:
Ok I'll work on that tonight with Alex

jkleiber reviewed (commented):
A few comments. I think we need to get the drivetrain PR merged. It would be good to see this PR tested in sim + advantage scope prior to merging as well

linglejack06 committed:
Diff drive (check if it can merge currently) (#30)

* Basic classes and constants

* add advantage kit kickoff release

* The advatange kit stuff has been added, probably doesn't work

* add getGyroRotation2d method to drive for interfacing with navigation class

- remove odometry from drive class since it is now in navigation
- fix Constants

* added Drive SimIO and working on drivecommands

* Basic classes and constants

* add advantage kit kickoff release

* The advatange kit stuff has been added, probably doesn't work

* add getGyroRotation2d method to drive for interfacing with navigation class

- remove odometry from drive class since it is now in navigation
- fix Constants

* Integrate vision/navigation into robot container (#23)

* integrate navigation into robot container

* add simulated pose input to drive and replace suppliers for navigation with real method calls

* fix unused import statements and switch to logged robot

* Ignore the ff stuff, maybe the diffdrive command works

* Fixed past merge error, it builds but idk if it works

* fixed the build gradle

* fixed constant and conatiner problems from the pull comments, not everything done

* add other two suppliers for navigation

* fix formatting errors

* move navigation and drive init to constructor

* format code

* move supplier variable init to outside constructor to keep private

* format again?

* working on controlDriveTrain(not sure it is correct at all)

* deleted useless code and added state controller

* made controlDriveTrain called in periodic and added helper method

* nothing

* add last est timestamp var

---------

Co-authored-by: Alexander Li <alexliblacksburg@gmail.com>

linglejack06 committed:
Merge branch 'main' into aim-at-target

linglejack06 committed:
remove unneccesary vars / imports

linglejack06 committed:
fix dot product

linglejack06 committed:
add method to determine alliance

linglejack06 committed:
add current heading getter

AlexEvergarden committed:
created setTarget heading

linglejack06 committed:
add command

linglejack06 commented:
just updated drive with @AlexEvergarden. according to what Charlie did. However it looks wrong as the pid controller takes in current heading and target heading (but target heading is constantly changing as the target heading method updates based on current position?)

linglejack06 committed:
remove stop

linglejack06 committed:
add dummy command to robot container

linglejack06 committed:
fix semicolon

linglejack06 committed:
add mode

linglejack06 committed:
update placeholders

jkleiber commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -15,14 +15,15 @@ public class Drive extends SubsystemBase {
-    private final SimpleMotorFeedforward driveff =
-            new SimpleMotorFeedforward(Constants.DriveConstants.kS, Constants.DriveConstants.kV);
+
+    private PIDController thetaController = new PIDController(0, 0, 0); // placeholders
```
This will have to have non-zero gains in order to control the robot to line up with the target

jkleiber commented on src/main/java/frc/robot/commands/AimAtTarget.java:
```diff
@@ -0,0 +1,29 @@
+    Navigation navSubsystem;
+    int targetId;
+
+    public AimAtTarget(Drive driveSubsystem, Navigation navSubsystem, int targetId) {
```
This command should have a `end()` method that puts the drivetrain back into the MANUAL state when the command ends

jkleiber commented on src/main/java/frc/robot/commands/AimAtTarget.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public void execute() {
+        driveSubsystem.aim(navSubsystem.getCurrentHeading(), navSubsystem.getTargetHeading());
```
My preference would be for us to have a generic `setAction` function that allows us to say we want to aim, and then create a supplier that feeds heading error to the Drive subssytem from Navigation.

Then in `controlDriveTrain` we would call the `aim` function in the `AIM` state, which would use the PID to drive the supplied error to 0 via rotation commands

jkleiber replied:
Note that setAction would not set the State - it would be used to transition between states. See the main robot shooter code for details

jkleiber reviewed (commented):
Some comments on the drivetrain side of things

linglejack06 committed:
add updated drive state and fix relevant commands + fix 0 pid

linglejack06 commented:
fixed majority of things, might need to update state still

jkleiber commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -18,8 +22,7 @@
 
-    public static enum driveTrainState {
-
+    public static enum DriveTrainState {
```
Thank you for this

jkleiber commented on src/main/java/frc/robot/commands/AimAtTarget.java:
```diff
@@ -0,0 +1,39 @@
+        driveSubsystem.setDriveState(DriveTrainState.MANUAL);
+    }
+
+    @Override
+    public boolean isFinished() {
+        return Math.abs(navSubsystem.getCurrentHeading() - navSubsystem.getTargetHeading()) < 1e-5;
+    }
```
This is good logic for determining if you are aligned, but you don't actually want to stop the command unless the aim button is released

linglejack06 replied:
ok ill remove it then.

jkleiber commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -49,6 +73,16 @@ public void stop() {
+                thetaController.calculate(
+                        currentHeading.getAsDouble(), targetHeading.getAsDouble());
+        if (Math.abs(currentHeading.getAsDouble() - targetHeading.getAsDouble()) < 1e-5) {
+            mode = DriveTrainState.MANUAL;
```
Let's stay in the AIM state even if the error is small, but supply some boolean that says we are aligned (this will be used by the shooter to determine if we are allowed to shoot)

jkleiber reviewed (commented):
I'm feeling pretty good about this change, just a couple of suggestions before we finalize this

linglejack06 committed:
add aligned to replace switch to manual

linglejack06 committed:
remove is finished logic

linglejack06 commented:
added requested changes

AlexEvergarden committed:
fully added advatange kit to robot.java, fixed some null values errors in visIOSim

AlexEvergarden committed:
trying to get the robot to turn in sim, by changing rbt cntnr, but not working

linglejack06 committed:
fix auto log output error

linglejack06 committed:
add reference pose and current pose method

AlexEvergarden committed:
deleted useless code

linglejack06 committed:
fix constants

linglejack06 committed:
add closed loop[ false

linglejack06 committed:
fix turn error

linglejack06 committed:
invert forward

AlexEvergarden committed:
inverted the turn

linglejack06 committed:
add autologged current pose

linglejack06 committed:
add auto log

linglejack06 committed:
update aim method to actually control drive

linglejack06 committed:
remove theta controller (aim at target works)

linglejack06 committed:
format code

linglejack06 commented:
aim at target now works in simulation

linglejack06 commented:
vision not working with simulation ( throws pose way off ).

linglejack06 committed:
prevent double vision pose from being added

linglejack06 committed:
add camera uncertainty to improce pose from 2024 code

linglejack06 committed:
make single tag uncertainty greater

linglejack06 commented:
i think this is ready for review. Only optimization needed is making robot not only turn counter-clockwise when aiming.

jkleiber commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -49,6 +73,15 @@ public void stop() {
 
+    public void aim() {
+        this.forward = 0;
+        this.rotation = targetHeading.getAsDouble();
```
The reason the robot is turning the same direction every time is because we are passing in the target heading into Arcade Drive directly.

It would be better to use a PID controller that changes this rotation variable based on the error between target heading and current heading. Make sure to enable continuous input in the PID

jkleiber commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -49,6 +73,15 @@ public void stop() {
+        this.forward = 0;
+        this.rotation = targetHeading.getAsDouble();
+        this.driveArcade(forward, rotation);
+        if (Math.abs(currentHeading.getAsDouble() - targetHeading.getAsDouble()) < 1e-5) {
```
This 1e-5 is a very very tight threshold. I would say aim for like 1 degree at first, otherwise we will never be aligned

jkleiber reviewed (commented):
Hopefully this helps with the turning thing

linglejack06 committed:
add pid controller + make aligned truth easier

linglejack06 commented:
added the pid controller, however, now it stops turning the bot after robot has gotten about halfway to the target heading. It then will go back the opposite way, like it believes the setpoint is halfway there. Do you have any clue how to fix this @jkleiber ?

jkleiber commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -49,6 +78,15 @@ public void stop() {
+        this.forward = 0;
+        this.rotation = rotationController.calculate(currentHeading.getAsDouble(), targetHeading.getAsDouble());
+        this.driveArcade(forward, rotation);
+        if (Math.abs(currentHeading.getAsDouble() - targetHeading.getAsDouble()) < 1) {
```
This is 1 radian, not 1 degree

jkleiber commented:
@linglejack06 To help debug the controller behavior, it would be good to look at a line graph plot in AdvantageScope of the target heading and the robot heading. The controller looks pretty correct to me at a glance, so if you look at the data / post it here maybe that will help us figure it out

linglejack06 committed:
vision pose more reliable now

linglejack06 committed:
update single tag deviation and remove pid controller

linglejack06 commented:
@jkleiber whats happening by looking at the line graph is the target heading is constantly moving closer to 0, as the robot gets closet to the target heading (the function returns the error). So at about halfway the current heading is around 1.5 and the error is still 1.5. When this happens the PID controller thinks the current heading is at the set point (target heading).

jkleiber commented:
It sounds like you're making the current heading aim for the heading error, let me check

jkleiber commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -49,6 +85,15 @@ public void stop() {
 
+    public void aim() {
+        this.forward = 0;
+        this.rotation = targetHeading.getAsDouble();
```
Here you would want the rotation controller that takes in a setpoint of 0 and a process variable of target heading

Naming this targetHeading is confusing because it's really headingError that you're representing

What you have here is fundamentally a P controller with gain of 1 (since your control output is simply the error itself). For an error in radians, you may need higher gain since a 0.1 radian error is large but will not result in much turn power

If it slows down early you may need to increase the P or I gain to get better results

linglejack06 replied:
@jkleiber. Just ran the PID Controller with the heading error as measurement and 0 as setpoint. It resulted in the bot spinning in circles (although it did slow down minutely around the setpoint). Here is a graph (yellow is heading error; blue is the robot heading).
<img width="459" alt="Screenshot 2024-02-14 at 6 01 41 PM" src="https://github.com/team401/2024-everybot/assets/108961470/bb25a019-48f2-4f7a-bd39-10c7586927a8">

linglejack06 replied:
I have fixed this by updating target calculation. Now it takes the target vector y and x and performs atan2. Without accounting for current heading, so it returns the actual target heading rather than error).

By adding this and updating pid controller to have process variable of current heading and set point as target heading, the controller works how it should. Now just need to update PID values to prevent over rotating.
updated picture:
<img width="195" alt="Screenshot 2024-02-14 at 6 31 50 PM" src="https://github.com/team401/2024-everybot/assets/108961470/96cf388d-6cf8-493c-94b3-8bfc3f289e44">

jkleiber replied:
Cool! Let me know when you need another review. This looks good though (I'm glad you're getting a chance to learn more about PID by messing with this as well)

jkleiber reviewed (commented):
Hopefully this is helpful. We can discuss on Saturday in person if you'll be there

linglejack06 committed:
update target heading calc

linglejack06 committed:
move hardcoded numbers to Constants file

linglejack06 committed:
update aligned conditional + increase std devs

linglejack06 committed:
refactor aim at target command to take in align state

- command then converts to target id based on alliance

linglejack06 committed:
update standard devs and add prev pose for conditional

linglejack06 commented:
I think this should be ready to merge. estimated pose is a lot more accurate now and aiming works accurately on each side.
